pkg/provider/xml/xml.go

```diff
@@ -148,50 +118,50 @@ func DecodeAttributeQuery(request string) (*samlp.AttributeQueryType, error) {
 }
 
 func DecodeLogoutRequest(encoding string, message string) (*samlp.LogoutRequestType, error) {
-	reqBytes, err := base64.StdEncoding.DecodeString(message)
+	data, err := InflateAndDecode(encoding, true, message)
 	if err != nil {
 		return nil, err
 	}
-
 	req := &samlp.LogoutRequestType{}
-	switch encoding {
-	case "":
-		reader := flate.NewReader(bytes.NewReader(reqBytes))
-		decoder := xml.NewDecoder(reader)
-		if err = decoder.Decode(req); err != nil {
-			return nil, err
-		}
-	case EncodingDeflate:
-		reader := flate.NewReader(bytes.NewReader(reqBytes))
-		decoder := xml.NewDecoder(reader)
-		if err = decoder.Decode(req); err != nil {
-			return nil, err
-		}
-	default:
-		return nil, fmt.Errorf("unknown encoding")
+	if err := xml.Unmarshal(data, req); err != nil {
+		fmt.Println(err.Error())
+		return nil, err
 	}
-
 	return req, nil
 }
 
-func DecodeResponse(encoding string, message string) (*samlp.ResponseType, error) {
-
+func DecodeResponse(encoding string, b64 bool, message string) (*samlp.ResponseType, error) {
+	data, err := InflateAndDecode(encoding, b64, message)
+	if err != nil {
+		return nil, err
+	}
 	req := &samlp.ResponseType{}
-	switch encoding {
-	case "":
-		decoder := xml.NewDecoder(bytes.NewReader([]byte(message)))
-		if err := decoder.Decode(req); err != nil {
+	if err := xml.Unmarshal(data, req); err != nil {
+		fmt.Println(err.Error())
+		return nil, err
+	}
+	return req, nil
+}
+
+func InflateAndDecode(encoding string, b64 bool, message string) (_ []byte, err error) {
+	data := []byte(message)
+	if b64 {
+		data, err = base64.StdEncoding.DecodeString(message)
+		if err != nil {
 			return nil, err
 		}
+	}
+	switch encoding {
+	case "":
+		return data, nil
 	case EncodingDeflate:
-		reader := flate.NewReader(bytes.NewReader([]byte(message)))
-		decoder := xml.NewDecoder(reader)
-		if err := decoder.Decode(req); err != nil {
+		r := flate.NewReader(bytes.NewBuffer(data))
+		out, err := io.ReadAll(r)
+		if err != nil {
 			return nil, err
 		}
-	default:
-		return nil, fmt.Errorf("unknown encoding")
+		r.Close()
+		return out, nil
 	}
-
-	return req, nil
+	return nil, fmt.Errorf("unknown encoding")
```

Comment:
I would change it to:
```go
switch encoding {
	case "":
		return data, nil
	case EncodingDeflate:
		r := flate.NewReader(bytes.NewBuffer(data))
		defer r.Close()
		return io.ReadAll(r)
	default:
		return nil, ErrUnknownEncoding
}
```

Reply:
True, changed it to the suggestion with defer.